File: crossword_generator.js
// crossword_generator.js — JavaScript версия

const fs = require('fs');

class MiniCrossword {
    constructor(size = 5, seed = null) {
        this.size = size;
        this.grid = Array.from({ length: size }, () => Array(size).fill(' '));
        this.solution = Array.from({ length: size }, () => Array(size).fill(' '));
        this.words = ['АРБУЗ', 'БУКЕТ', 'ГАМА', 'ДОМ', 'ЕЛЬ', 'ЁЖИК', 'ЖУК', 'ЗОНТ', 'ИГЛА'];
        this.placedWords = [];
        this.wordPositions = {};
        if (seed !== null) {
            this._seed = seed;
        }
    }

    _rand() {
        if (this._seed !== undefined) {
            this._seed = (this._seed * 9301 + 49297) % 233280;
            return this._seed / 233280;
        }
        return Math.random();
    }

    _shuffle(arr) {
        for (let i = arr.length - 1; i > 0; i--) {
            const j = Math.floor(this._rand() * (i + 1));
            [arr[i], arr[j]] = [arr[j], arr[i]];
        }
        return arr;
    }

    generate() {
        const sortedWords = [...this.words].sort((a, b) => b.length - a.length);
        for (const word of sortedWords) {
            if (this._placeWord(word)) {
                this.placedWords.push(word);
            }
        }

        for (let i = 0; i < this.size; i++) {
            for (let j = 0; j < this.size; j++) {
                this.solution[i][j] = this.grid[i][j];
            }
        }

        const letters = 'АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ';
        for (let i = 0; i < this.size; i++) {
            for (let j = 0; j < this.size; j++) {
                if (this.grid[i][j] === ' ') {
                    this.grid[i][j] = letters[Math.floor(this._rand() * letters.length)];
                }
            }
        }
        return this.grid;
    }

    _placeWord(word) {
        const dirs = [[0, 1], [1, 0]];
        for (let attempt = 0; attempt < 50; attempt++) {
            const row = Math.floor(this._rand() * this.size);
            const col = Math.floor(this._rand() * this.size);
            const dir = dirs[Math.floor(this._rand() * dirs.length)];
            if (this._canPlace(word, row, col, dir)) {
                this._placeWordInGrid(word, row, col, dir);
                this.wordPositions[word] = [row, col, dir];
                return true;
            }
        }
        return false;
    }

    _canPlace(word, row, col, dir) {
        const [dr, dc] = dir;
        if (row + dr * (word.length - 1) >= this.size || col + dc * (word.length - 1) >= this.size) {
            return false;
        }
        for (let i = 0; i < word.length; i++) {
            const r = row + dr * i;
            const c = col + dc * i;
            if (this.grid[r][c] !== ' ' && this.grid[r][c] !== word[i]) {
                return false;
            }
        }
        return true;
    }

    _placeWordInGrid(word, row, col, dir) {
        const [dr, dc] = dir;
        for (let i = 0; i < word.length; i++) {
            const r = row + dr * i;
            const c = col + dc * i;
            this.grid[r][c] = word[i];
        }
    }

    _getNumber(row, col) {
        let num = 1;
        for (const word of Object.keys(this.wordPositions)) {
            const [r, c, dir] = this.wordPositions[word];
            if (r === row && c === col) {
                return num;
            }
            num++;
        }
        return 0;
    }

    printCrossword(showSolution = false) {
        const grid = showSolution ? this.solution : this.grid;
        console.log('┌─────┬─────┬─────┬─────┬─────┐');
        for (let i = 0; i < this.size; i++) {
            let line = '│';
            for (let j = 0; j < this.size; j++) {
                if (grid[i][j] === ' ') {
                    line += '   ';
                } else {
                    const num = this._getNumber(i, j);
                    if (num > 0 && !showSolution) {
                        line += `${String(num).padStart(2)} `;
                    } else {
                        line += ` ${grid[i][j]} `;
                    }
                }
                if (j < this.size - 1) line += '│';
            }
            line += '│';
            console.log(line);
            if (i < this.size - 1) {
                console.log('├─────┼─────┼─────┼─────┼─────┤');
            }
        }
        console.log('└─────┴─────┴─────┴─────┴─────┘');
    }

    printClues() {
        const words = Object.keys(this.wordPositions);
        console.log('\nСлова по горизонтали:');
        for (const word of words) {
            const [r, c, dir] = this.wordPositions[word];
            if (dir[0] === 0 && dir[1] === 1) {
                console.log(`${words.indexOf(word) + 1}. ${word}`);
            }
        }
        console.log('\nСлова по вертикали:');
        for (const word of words) {
            const [r, c, dir] = this.wordPositions[word];
            if (dir[0] === 1 && dir[1] === 0) {
                console.log(`${words.indexOf(word) + 1}. ${word}`);
            }
        }
    }

    saveJSON(filename = 'crossword.json') {
        const data = {
            size: this.size,
            grid: this.grid,
            solution: this.solution,
            words: this.placedWords,
            positions: this.wordPositions
        };
        fs.writeFileSync(filename, JSON.stringify(data, null, 2));
        console.log(`💾 Сохранено: ${filename}`);
    }

    saveHTML(filename = 'crossword.html') {
        let html = `<!DOCTYPE html>
<html>
<head><meta charset="UTF-8"><title>Кроссворд</title>
<style>
body { font-family: monospace; background: #1a1a2e; color: #fff; padding: 20px; }
table { border-collapse: collapse; margin: 20px auto; }
td { width: 40px; height: 40px; text-align: center; font-size: 18px; border: 2px solid #444; background: #2a2a4e; }
td.filled { background: #3a3a6e; }
td.empty { background: #1a1a2e; }
.number { font-size: 10px; color: #888; position: relative; top: -15px; left: -15px; }
</style>
</head>
<body>
<h1 style="text-align:center;">🧩 Кроссворд</h1>
<table>
`;
        for (let i = 0; i < this.size; i++) {
            html += '<tr>';
            for (let j = 0; j < this.size; j++) {
                const cell = this.grid[i][j];
                const cls = cell === ' ' ? 'empty' : 'filled';
                const num = this._getNumber(i, j);
                const display = cell === ' ' ? '&nbsp;' : cell;
                const numHtml = num > 0 ? `<span class="number">${num}</span>` : '';
                html += `<td class="${cls}">${numHtml}${display}</td>`;
            }
            html += '</tr>';
        }
        html += `</table>
<p style="text-align:center;">💡 Подсказки доступны в консоли</p>
</body></html>`;
        fs.writeFileSync(filename, html);
        console.log(`💾 Сохранено: ${filename}`);
    }
}

function main() {
    const readline = require('readline').createInterface({
        input: process.stdin,
        output: process.stdout
    });

    console.log('🧩 Mini Crossword Generator (JavaScript)');

    readline.question('Размер кроссворда (5 или 7): ', (sizeInput) => {
        let size = 5;
        if (sizeInput && [5, 7].includes(parseInt(sizeInput))) {
            size = parseInt(sizeInput);
        }

        readline.question('Seed (число, опционально): ', (seedInput) => {
            const seed = seedInput ? parseInt(seedInput) : null;

            const gen = new MiniCrossword(size, seed);
            gen.generate();

            console.log('\nКроссворд:');
            gen.printCrossword();

            readline.question('\nПоказать решение? (y/n): ', (show) => {
                if (show.toLowerCase() === 'y') {
                    console.log('\nРешение:');
                    gen.printCrossword(true);
                }

                gen.printClues();
                gen.saveJSON();
                gen.saveHTML();
                readline.close();
            });
        });
    });
}

if (require.main === module) main();

File: crossword_generator.rb
# crossword_generator.rb — Ruby версия

class MiniCrossword
  attr_reader :grid, :solution, :placed_words, :word_positions

  def initialize(size = 5, seed = nil)
    @size = size
    @grid = Array.new(size) { Array.new(size, ' ') }
    @solution = Array.new(size) { Array.new(size, ' ') }
    @words = ['АРБУЗ', 'БУКЕТ', 'ГАМА', 'ДОМ', 'ЕЛЬ', 'ЁЖИК', 'ЖУК', 'ЗОНТ', 'ИГЛА']
    @placed_words = []
    @word_positions = {}
    @rng = seed ? Random.new(seed) : Random.new
  end

  def generate
    sorted_words = @words.sort_by { |w| -w.length }

    sorted_words.each do |word|
      if place_word(word)
        @placed_words << word
      end
    end

    @size.times do |i|
      @size.times do |j|
        @solution[i][j] = @grid[i][j]
      end
    end

    letters = 'АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ'.chars
    @size.times do |i|
      @size.times do |j|
        @grid[i][j] = letters.sample(random: @rng) if @grid[i][j] == ' '
      end
    end
  end

  def place_word(word)
    dirs = [[0, 1], [1, 0]]
    50.times do
      row = @rng.rand(@size)
      col = @rng.rand(@size)
      dir = dirs.sample(random: @rng)
      if can_place?(word, row, col, dir)
        place_word_in_grid(word, row, col, dir)
        @word_positions[word] = [row, col, dir]
        return true
      end
    end
    false
  end

  def can_place?(word, row, col, dir)
    dr, dc = dir
    return false if row + dr * (word.length - 1) >= @size || col + dc * (word.length - 1) >= @size
    word.chars.each_with_index do |ch, i|
      r = row + dr * i
      c = col + dc * i
      return false if @grid[r][c] != ' ' && @grid[r][c] != ch
    end
    true
  end

  def place_word_in_grid(word, row, col, dir)
    dr, dc = dir
    word.chars.each_with_index do |ch, i|
      r = row + dr * i
      c = col + dc * i
      @grid[r][c] = ch
    end
  end

  def get_number(row, col)
    num = 1
    @word_positions.each do |_, (r, c, _)|
      return num if r == row && c == col
      num += 1
    end
    0
  end

  def print_crossword(show_solution = false)
    grid = show_solution ? @solution : @grid
    puts "┌─────┬─────┬─────┬─────┬─────┐"
    @size.times do |i|
      print "│"
      @size.times do |j|
        if grid[i][j] == ' '
          print "   "
        else
          num = get_number(i, j)
          if num > 0 && !show_solution
            print "#{num.to_s.rjust(2)} "
          else
            print " #{grid[i][j]} "
          end
        end
        print "│" if j < @size - 1
      end
      puts "│"
      puts "├─────┼─────┼─────┼─────┼─────┤" if i < @size - 1
    end
    puts "└─────┴─────┴─────┴─────┴─────┘"
  end

  def print_clues
    puts "\nСлова по горизонтали:"
    @placed_words.each_with_index do |word, i|
      if @word_positions[word][2] == [0, 1]
        puts "#{i + 1}. #{word}"
      end
    end
    puts "\nСлова по вертикали:"
    @placed_words.each_with_index do |word, i|
      if @word_positions[word][2] == [1, 0]
        puts "#{i + 1}. #{word}"
      end
    end
  end

  def save_json(filename = 'crossword.json')
    data = {
      size: @size,
      grid: @grid,
      solution: @solution,
      words: @placed_words
    }
    File.write(filename, JSON.pretty_generate(data))
    puts "💾 Сохранено: #{filename}"
  end

  def save_html(filename = 'crossword.html')
    html = <<~HTML
      <!DOCTYPE html>
      <html>
      <head><meta charset="UTF-8"><title>Кроссворд</title>
      <style>
      body { font-family: monospace; background: #1a1a2e; color: #fff; padding: 20px; }
      table { border-collapse: collapse; margin: 20px auto; }
      td { width: 40px; height: 40px; text-align: center; font-size: 18px; border: 2px solid #444; background: #2a2a4e; }
      td.filled { background: #3a3a6e; }
      td.empty { background: #1a1a2e; }
      .number { font-size: 10px; color: #888; position: relative; top: -15px; left: -15px; }
      </style>
      </head>
      <body>
      <h1 style="text-align:center;">🧩 Кроссворд</h1>
      <table>
    HTML
    @size.times do |i|
      html << "<tr>"
      @size.times do |j|
        cell = @grid[i][j]
        cls = cell == ' ' ? 'empty' : 'filled'
        num = get_number(i, j)
        display = cell == ' ' ? '&nbsp;' : cell
        num_html = num > 0 ? "<span class=\"number\">#{num}</span>" : ''
        html << "<td class=\"#{cls}\">#{num_html}#{display}</td>"
      end
      html << "</tr>\n"
    end
    html << <<~HTML
      </table>
      <p style="text-align:center;">💡 Подсказки доступны в консоли</p>
      </body>
      </html>
    HTML
    File.write(filename, html)
    puts "💾 Сохранено: #{filename}"
  end
end

def main
  puts "🧩 Mini Crossword Generator (Ruby)"
  print "Размер кроссворда (5 или 7): "
  size_input = gets.chomp
  size = size_input.empty? ? 5 : size_input.to_i
  size = 5 unless [5, 7].include?(size)

  print "Seed (число, опционально): "
  seed_input = gets.chomp
  seed = seed_input.empty? ? nil : seed_input.to_i

  gen = MiniCrossword.new(size, seed)
  gen.generate

  puts "\nКроссворд:"
  gen.print_crossword

  print "\nПоказать решение? (y/n): "
  show = gets.chomp
  if show.downcase == 'y'
    puts "\nРешение:"
    gen.print_crossword(true)
  end

  gen.print_clues
  gen.save_json
  gen.save_html
end

main if __FILE__ == $0
